Support NeXus `default_slice` attribute

axelboc opened the pull request:
Fix #1844 

[Screencast from 2025-09-16 15-57-57.webm](https://github.com/user-attachments/assets/730f8d2e-ef16-4d6c-8927-db5772f11e8d)

---

Note that the validation of the `default_slice` is done in two stages:

- First when retrieving the attribute from the `NXdata` group (`getDefaultSlice`) — at this stage, I check that the value of the attribute is an array of strings, that the array has the same size of the signal's dimensions, and that every value is either `'.'` or a valid number within the bounds of the corresponding dimension.
- Second when actually applying the default slice (`applyDefaultSlice`) — at this stage, I check that the number of `'.'` in the default slice arrays corresponds to the number of "mappable" dimensions (i.e. the number of axes + the number of locked dimensions for the selected visualization)

If at either stage the default slice doesn't pass the validity checks, it is simply ignored.

axelboc commented:
/approve

axelboc commented on cypress/snapshots/app.cy.ts/default_slice.snap.png:
New screenshot here (as usual, the others are just diffs because of the new `NXdata` group in the explorer)

loichuder reviewed (commented):
Perhaps I missed it but should we not print warnings in the console if the `default_slice` is malformed?

I think that if the `default_slice` attribute is present, there is a will from the file creator to use this feature so that any issue should probably be raised to understand what is going on

axelboc commented:
Sorry, I just forgot about it!

I've now added two warnings:

- when the length doesn't match the number of dimensions
- when the indices are out of bounds (or NaN)

Like with `SILX_style`, I do not warn if the shape/type of the attribute is not what we expect (i.e. an array of strings), as this seems highly unlikely to happen. I've actually also removed the mock `NXdata` group that used to test `default_slice` with a mixed array of strings and numbers, as I doubt anyone would create a `default_slice` attribute with a compound type in practice... :smile:

axelboc commented:
/approve

axelboc commented:
/approve

axelboc commented on packages/app/src/vis-packs/nexus/utils.ts:
```diff
@@ -231,6 +237,48 @@ export function findTitleDataset(
+
+  if (!isValid) {
+    // eslint-disable-next-line no-console
+    console.warn(
```
I'm not fussed to be honest. It adds a little bit of complexity for very little gain. Same for the other warning. As discussed, H5Web is not a NeXus format validator and users have to know to open their dev tools anyway, so very few are likely to see these messages... Informing that the `default_slice` attribute has been considered but rejected, and roughly why, is good enough in my opinion. If they really don't understand why from the warning, they can always open an issue.

loichuder replied:
> users have to know to open their dev tools anyway, so very few are likely to see these messages...

Fair
